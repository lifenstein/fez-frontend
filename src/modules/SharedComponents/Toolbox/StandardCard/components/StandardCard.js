import React from 'react';
import PropTypes from 'prop-types';
import { styled } from '@mui/material/styles';

import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import { HelpIcon } from '../../HelpDrawer';

const StyledCard = styled(Card)(({ theme }) => ({
    overflow: 'unset',
    fontWeight: theme.typography.fontWeightRegular,
}));
const StyledCardHeader = styled(CardHeader)(({ theme }) => ({
    borderRadius: '4px 4px 0px 0px',
    padding: '12px 24px',
    '& .MuiCardHeader-title': {
        [theme.breakpoints.down('sm')]: {
            ...theme.typography.h6,
        },
    },
}));

export const StandardCard = ({
    accentHeader,
    children,
    customBackgroundColor,
    customTitleBgColor,
    customTitleColor,
    fullHeight,
    help,
    noHeader,
    noPadding,
    primaryHeader,
    smallTitle = false,
    squareTop,
    standardCardId,
    subCard = false,
    title,
}) => {
    const customBG = !!customBackgroundColor ? { backgroundColor: customBackgroundColor } : null;
    const customTitleBG = !!customTitleBgColor ? { backgroundColor: customTitleBgColor } : null;
    const customTitle = !!customTitleColor ? { color: customTitleColor } : null;
    const fullHeightActual = !!fullHeight ? { height: '100%' } : null;
    const squareTopActual = !!squareTop ? { borderTopLeftRadius: 0, borderTopRightRadius: 0 } : null;
    const standardCardIdActual = !!standardCardId
        ? standardCardId
        : `standard-card${typeof title === 'string' ? '-' + title.replace(/ /g, '-').toLowerCase() : ''}`;

    return (
        <StyledCard
            id={standardCardIdActual}
            data-testid={standardCardIdActual}
            className={'StandardCard'}
            sx={{ ...customBG, ...customTitle, ...fullHeightActual }}
        >
            {!noHeader && (
                <StyledCardHeader
                    sx={{
                        ...(accentHeader || primaryHeader
                            ? { color: 'white.main', backgroundColor: `${primaryHeader ? 'primary' : 'accent'}.main` }
                            : {}),
                        ...squareTopActual,
                        ...customTitleBG,
                    }}
                    title={title}
                    titleTypographyProps={{
                        variant: smallTitle ? 'h6' : 'h5',
                        component: subCard ? 'h4' : 'h3',
                        color: 'inherit',
                        'data-testid': `${standardCardIdActual}-header`,
                        id: `${standardCardIdActual}-header`,
                    }}
                    action={!!help && !!help.text && <HelpIcon {...help} />}
                />
            )}
            <CardContent
                id={`${standardCardIdActual}-content`}
                data-analyticsid={`${standardCardIdActual}-content`}
                data-testid={`${standardCardIdActual}-content`}
                sx={{ ...(noPadding ? { padding: 0 } : {}) }}
            >
                {children}
            </CardContent>
        </StyledCard>
    );
};

StandardCard.propTypes = {
    accentHeader: PropTypes.bool,
    children: PropTypes.any,
    customBackgroundColor: PropTypes.any,
    customTitleBgColor: PropTypes.any,
    customTitleColor: PropTypes.any,
    fullHeight: PropTypes.bool,
    help: PropTypes.object,
    noHeader: PropTypes.bool,
    noPadding: PropTypes.bool,
    primaryHeader: PropTypes.bool,
    smallTitle: PropTypes.bool,
    squareTop: PropTypes.bool,
    standardCardId: PropTypes.string,
    subCard: PropTypes.bool,
    title: PropTypes.any,
};

export default StandardCard;
