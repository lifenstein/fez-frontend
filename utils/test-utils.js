/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');
const mime = require('mime-types');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

export const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }), renderMethod = render } = {},
) => {
    return renderMethod(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const WithRouter = ({ children, route = '/', history = createMemoryHistory({ initialEntries: [route] }) }) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

export const defaultDatastream = {
    dsi_embargo_date: null,
    dsi_open_access: 1,
    dsi_label: '',
    dsi_copyright: null,
    dsi_state: 'A',
    dsi_size: 27932352,
};
export const createFezDatastreamInfoArray = (datastreams, pid = null, withPreview = true) => {
    let processed = datastreams.filter(datastream => typeof datastream === 'object' || typeof datastream === 'string');
    if (withPreview) {
        processed = [];
        datastreams.forEach(function(datastream) {
            processed.push(datastream);
            const filename = typeof datastream === 'object' ? datastream.dsi_dsid : datastream;
            const mimetype = mime.lookup(filename);
            // bail in case it's not derivable
            if (
                !filename.includes('.pdf') &&
                !mimetype.includes('image') &&
                !mimetype.includes('audio') &&
                !mimetype.includes('video')
            ) {
                return;
            }
            // eslint-disable-next-line no-nested-ternary
            const derivativeMimetype = mimetype.includes('audio')
                ? 'audio/mp3'
                : mimetype.includes('video')
                ? 'video/mp4'
                : 'image/jpg';

            (derivativeMimetype.includes('image') ? ['preview_', 'thumbnail_', 'web_'] : ['']).forEach(function(
                prefix,
            ) {
                processed.push({
                    dsi_dsid: `${prefix}${filename.split('.')[0]}_t.${derivativeMimetype.split('/').pop()}`,
                    dsi_mimetype: derivativeMimetype,
                });
            });
        });
    }

    return processed.map((datastream, index) => ({
        ...defaultDatastream,
        ...{
            dsi_pid: pid,
            dsi_order: index + 1,
            dsi_mimetype: mime.lookup(typeof datastream === 'object' ? datastream.dsi_dsid : datastream),
        },
        ...(typeof datastream === 'object'
            ? datastream
            : {
                  dsi_dsid: datastream,
              }),
    }));
};

export const withDatastreams = (sources, datastreams, callback) =>
    datastreams.map(item => {
        const source = sources.find(datastream =>
            datastream.dsi_dsid.includes(
                item.dsi_dsid
                    .replace('_t.', '.')
                    .replace('preview_', '')
                    .replace('thumbnail_', '')
                    .replace('web_', '')
                    .split('.')[0],
            ),
        );
        const isDerivative = item.dsi_dsid !== source.dsi_dsid;
        callback(item, source, isDerivative);
    });

export const getDatastreamByFilename = (filename, derivatives) =>
    derivatives.find(derivative => derivative.dsi_dsid === filename);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    AllTheProviders,
    WithReduxStore,
    WithRouter,
    createFezDatastreamInfoArray,
    getDatastreamByFilename,
    withDatastreams,
};
