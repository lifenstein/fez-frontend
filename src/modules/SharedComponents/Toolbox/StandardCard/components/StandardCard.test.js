import React from 'react';
import { render, WithReduxStore } from 'test-utils';

import StandardCard from './StandardCard';

jest.mock('@mui/styles', () => ({
    makeStyles: () => () => ({}),
}));

function setup(testProps) {
    const props = {
        title: 'card title',
        classes: {
            card: 'testClass',
        },
        ...testProps,
    };
    return render(
        <WithReduxStore>
            <StandardCard {...props} />
        </WithReduxStore>,
    );
}

describe('StandardCard component', () => {
    it('renders with title and no help icon', () => {
        const { asFragment } = setup({});
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with no title', () => {
        const { asFragment } = setup({ title: '' });
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with title and help button', () => {
        const { asFragment } = setup({
            help: {
                title: 'help',
                text: 'help text',
                buttonLabel: 'OK',
            },
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with custom colours and full height', () => {
        const { asFragment } = setup({
            customBackgroundColor: '#fcc',
            customTitleColor: '#111',
            customTitleBgColor: '#ccc',
            fullHeight: true,
            noPadding: true,
            primaryHeader: true,
            classes: {
                cardHeaderPurple: 'purple',
                cardContentNoPadding: 'no-padding',
                cardHeaderPrimary: '#555',
            },
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with square top and accent header', () => {
        const { asFragment } = setup({
            squareTop: true,
            accentHeader: true,
            classes: {
                cardHeaderAccent: '#333',
            },
        });

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with small title and as a subcard', () => {
        const { asFragment } = setup({
            smallTitle: true,
            subCard: true,
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders given ID for standard card', () => {
        const { asFragment } = setup({ standardCardId: 'test-card' });
        expect(asFragment()).toMatchSnapshot();
    });
});
