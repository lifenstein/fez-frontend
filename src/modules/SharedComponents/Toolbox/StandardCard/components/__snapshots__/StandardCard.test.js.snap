// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardCard component renders given ID for standard card 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-38 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="test-card"
    id="test-card"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-43 MuiTypography-h5 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="test-card-header"
          id="test-card-header"
        >
          card title
        </h3>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
      data-testid="test-card-content"
      id="test-card-content"
    />
  </div>
</DocumentFragment>
`;

exports[`StandardCard component renders with custom colours and full height 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-20 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standard-card-card-title"
    id="standard-card-card-title"
    style="background-color: rgb(255, 204, 204); color: rgb(17, 17, 17); height: 100%;"
  >
    <div
      class="MuiCardHeader-root makeStyles-cardHeaderPrimary-22"
      style="background-color: rgb(204, 204, 204);"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-25 MuiTypography-h5 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="standard-card-card-title-header"
          id="standard-card-card-title-header"
        >
          card title
        </h3>
      </div>
    </div>
    <div
      class="MuiCardContent-root makeStyles-cardContentNoPadding-21"
      data-testid="standard-card-card-title-content"
      id="standard-card-card-title-content"
    />
  </div>
</DocumentFragment>
`;

exports[`StandardCard component renders with no title 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-7 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standard-card-"
    id="standard-card-"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-12 MuiTypography-h5 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="standard-card--header"
          id="standard-card--header"
        />
      </div>
    </div>
    <div
      class="MuiCardContent-root"
      data-testid="standard-card--content"
      id="standard-card--content"
    />
  </div>
</DocumentFragment>
`;

exports[`StandardCard component renders with small title and as a subcard 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-32 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standard-card-card-title"
    id="standard-card-card-title"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h4
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-37 MuiTypography-h6 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="standard-card-card-title-header"
          id="standard-card-card-title-header"
        >
          card title
        </h4>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
      data-testid="standard-card-card-title-content"
      id="standard-card-card-title-content"
    />
  </div>
</DocumentFragment>
`;

exports[`StandardCard component renders with square top and accent header 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-26 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standard-card-card-title"
    id="standard-card-card-title"
  >
    <div
      class="MuiCardHeader-root makeStyles-cardHeaderAccent-29"
      style="border-top-left-radius: 0; border-top-right-radius: 0;"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-31 MuiTypography-h5 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="standard-card-card-title-header"
          id="standard-card-card-title-header"
        >
          card title
        </h3>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
      data-testid="standard-card-card-title-content"
      id="standard-card-card-title-content"
    />
  </div>
</DocumentFragment>
`;

exports[`StandardCard component renders with title and help button 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-13 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standard-card-card-title"
    id="standard-card-card-title"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-18 MuiTypography-h5 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="standard-card-card-title-header"
          id="standard-card-card-title-header"
        >
          card title
        </h3>
      </div>
      <div
        class="MuiCardHeader-action"
      >
        <button
          aria-label="Click for more information"
          class="MuiButtonBase-root MuiIconButton-root"
          data-testid="help-icon"
          id="help-icon"
          tabindex="0"
          title="Click for more information"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              class="MuiSvgIcon-root makeStyles-helpIcon-19"
              focusable="false"
              role="img"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
              />
              <title>
                Click for more information
              </title>
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
      data-testid="standard-card-card-title-content"
      id="standard-card-card-title-content"
    />
  </div>
</DocumentFragment>
`;

exports[`StandardCard component renders with title and no help icon 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-1 StandardCard MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="standard-card-card-title"
    id="standard-card-card-title"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiCardHeader-title makeStyles-responsiveHeader-6 MuiTypography-h5 MuiTypography-colorInherit MuiTypography-displayBlock"
          data-testid="standard-card-card-title-header"
          id="standard-card-card-title-header"
        >
          card title
        </h3>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
      data-testid="standard-card-card-title-content"
      id="standard-card-card-title-content"
    />
  </div>
</DocumentFragment>
`;
